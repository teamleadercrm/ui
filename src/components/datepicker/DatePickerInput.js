import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Box, { pickBoxProps } from '../box';
import DatePicker from '../datepicker';
import Icon from '../icon';
import Input from '../input';
import Popover from '../popover';
import cx from 'classnames';
import theme from './theme.css';
import { formatDate } from './localeUtils';
import { IconCalendarSmallOutline } from '@teamleader/ui-icons';

/** @type {React.ComponentType<any>} */
class DatePickerInput extends PureComponent {
  state = {
    isPopoverActive: false,
    popoverAnchorEl: null,
    selectedDate: null,
  };

  static getDerivedStateFromProps(props, state) {
    if (props.selectedDate !== undefined && props.selectedDate !== state.selectedDate) {
      return {
        selectedDate: props.selectedDate,
      };
    }

    return null;
  }

  getFormattedDate = () => {
    const { formatDate: customFormatDate, locale } = this.props;
    const { selectedDate } = this.state;

    if (!selectedDate) {
      return '';
    }

    if (!customFormatDate) {
      return formatDate(selectedDate, locale);
    }

    return customFormatDate(selectedDate, locale);
  };

  handleInputFocus = (event) => {
    const { onFocus, readOnly } = this.props.inputProps;
    const { openPickerOnFocus } = this.props;

    if (readOnly) {
      return;
    }

    if (openPickerOnFocus) {
      this.setState(
        {
          popoverAnchorEl: event.currentTarget,
          isPopoverActive: true,
        },
        () => onFocus && onFocus(event),
      );
    } else {
      onFocus && onFocus(event);
    }
  };

  handleInputClick = (event) => {
    const { onFocus, onClick } = this.props.inputProps;
    const { openPickerOnFocus } = this.props;

    if (!openPickerOnFocus) {
      this.setState(
        {
          popoverAnchorEl: event.currentTarget,
          isPopoverActive: true,
        },
        () => onFocus && onFocus(),
      );
    }

    onClick && onClick(event);
  };

  handlePopoverClose = () => {
    const { onBlur } = this.props;
    onBlur && onBlur({ relatedTarget: null });
    this.setState({ isPopoverActive: false });
  };

  handleDatePickerDateChange = (date) => {
    this.setState({ isPopoverActive: false, selectedDay: date }, () => this.props.onChange(date));
  };

  renderIcon = () => {
    const inverse = this.props.inputProps && this.props.inputProps.inverse;

    return (
      <Icon color={inverse ? 'teal' : 'neutral'} tint={inverse ? 'light' : 'darkest'} marginRight={2}>
        <IconCalendarSmallOutline />
      </Icon>
    );
  };

  render() {
    const {
      className,
      dayPickerProps,
      footer,
      inverse,
      inputProps,
      locale,
      popoverProps,
      size,
      inputSize,
      datePickerSize,
      ...others
    } = this.props;
    const { isPopoverActive, popoverAnchorEl, selectedDate } = this.state;
    const boxProps = pickBoxProps(others);
    const datePickerClassNames = cx(theme['date-picker-input'], theme[`is-${datePickerSize || size}`]);

    return (
      <Box className={className} {...boxProps}>
        <Input
          inverse={inverse}
          prefix={this.renderIcon()}
          size={inputSize || size}
          value={this.getFormattedDate()}
          width="120px"
          noInputStyling={dayPickerProps && dayPickerProps.withMonthPicker}
          {...inputProps}
          onClick={this.handleInputClick}
          onFocus={this.handleInputFocus}
        />
        <Popover
          active={isPopoverActive}
          anchorEl={popoverAnchorEl}
          backdrop="transparent"
          maxWidth="min-content"
          onEscKeyDown={this.handlePopoverClose}
          onOverlayClick={this.handlePopoverClose}
          position="end"
          offsetCorrection={30}
          returnFocusToSource={false}
          {...popoverProps}
        >
          <Box overflowY="auto">
            <DatePicker
              className={datePickerClassNames}
              locale={locale}
              month={selectedDate}
              onChange={this.handleDatePickerDateChange}
              selectedDate={selectedDate}
              size={datePickerSize || size}
              {...dayPickerProps}
            />
          </Box>
          {footer && (
            <Box
              backgroundColor="neutral"
              backgroundTint="light"
              borderColor="neutral"
              borderTint="normal"
              borderTopWidth={1}
              flex="1 0 auto"
              overflowX="hidden"
              paddingHorizontal={3}
              paddingVertical={3}
              tabIndex="0"
            >
              {footer}
            </Box>
          )}
        </Popover>
      </Box>
    );
  }
}

DatePickerInput.propTypes = {
  /** A class name for the wrapper to give custom styles. */
  className: PropTypes.string,
  /** Object with props for the DatePicker component. */
  dayPickerProps: PropTypes.object,
  /** A footer component, rendered at the bottom of the date picker */
  footer: PropTypes.any,
  /** A custom function to format a date. */
  formatDate: PropTypes.func,
  /** Object with props for the Input component. */
  inputProps: PropTypes.object,
  /** If true, component will be rendered in inverse mode. */
  inverse: PropTypes.bool,
  /** The language ISO locale code ('en-GB', 'nl-BE', 'fr-FR',...). */
  locale: PropTypes.string,
  /** Callback function that is fired when the date has changed. */
  onChange: PropTypes.func,
  /** Callback function that is fired when the popover with the calendar gets closed (unfocused) */
  onBlur: PropTypes.func,
  /** Object with props for the Popover component. */
  popoverProps: PropTypes.object,
  /** The current selected date. */
  selectedDate: PropTypes.instanceOf(Date),
  /** Size of the Input & DatePicker components. */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /** Overridable size of the Input component. */
  inputSize: PropTypes.oneOf(['small', 'medium', 'large']),
  /** Overridable size of the DatePicker component. */
  datePickerSize: PropTypes.oneOf(['small', 'medium', 'large']),
  /** Whether the picker should automatically open on input focus. True by default. */
  openPickerOnFocus: PropTypes.bool,
};

DatePickerInput.defaultProps = {
  dayPickerProps: {},
  inputProps: {},
  inverse: false,
  locale: 'en-GB',
  popoverProps: {},
  size: 'medium',
  openPickerOnFocus: true,
};

export default DatePickerInput;
