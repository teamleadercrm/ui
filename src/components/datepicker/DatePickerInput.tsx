import { IconCalendarSmallOutline, IconCloseBadgedSmallFilled } from '@teamleader/ui-icons';
import React, { ReactNode, useEffect, useState } from 'react';
import { DayPickerProps as ReactDayPickerProps, Modifier } from 'react-day-picker';
import DatePicker from '.';
import { SIZES } from '../../constants';
import Box, { pickBoxProps } from '../box';
import { BoxProps } from '../box/Box';
import Icon from '../icon';
import Input from '../input';
import { InputProps } from '../input/Input';
import Popover from '../popover';
import { PopoverProps } from '../popover/Popover';
import { formatDate, parseMultiFormatsDate } from './localeUtils';
import theme from './theme.css';
import { isAllowedDate } from './utils';

const DEFAULT_FORMAT = 'dd/MM/yyyy';
const ALLOWED_DATE_FORMATS = [DEFAULT_FORMAT, 'd/M/yyyy', 'dd.MM.yyyy', 'd.M.yyyy', 'dd-MM-yyyy', 'd-M-yyyy'];

export interface DatePickerInputProps<IsTypeable extends boolean = true> extends Omit<BoxProps, 'size' | 'onChange'> {
  /** A class name for the wrapper to give custom styles. */
  className?: string;
  /** Object with props for the DatePicker component. */
  dayPickerProps?: DayPickerProps;
  /** A footer component, rendered at the bottom of the date picker */
  footer?: ReactNode;
  /** A custom function to format a date if input is not typeable */
  formatDate?: IsTypeable extends true ? never : (selectedDate: Date, locale: string) => string;
  /** Object with props for the Input component. */
  inputProps?: InputProps;
  /** If true, component will be rendered in inverse mode. */
  inverse?: boolean;
  /** The language ISO locale code ('en-GB', 'nl-BE', 'fr-FR',...). */
  locale?: string;
  /** Callback function that is fired when the date has changed. */
  onChange?: (selectedDate: Date | undefined) => void;
  /** Callback function that is fired when the popover with the calendar gets closed (unfocused) */
  onBlur?: () => void;
  /** Object with props for the Popover component. */
  popoverProps?: PopoverProps;
  /** The current selected date. */
  selectedDate?: Date;
  /** Size of the Input & DatePicker components. */
  size?: Exclude<typeof SIZES[number], 'tiny' | 'smallest' | 'hero' | 'fullscreen'>;
  /** Overridable size of the Input component. */
  inputSize?: Exclude<typeof SIZES[number], 'tiny' | 'smallest' | 'hero' | 'fullscreen'>;
  /** Overridable size of the DatePicker component. */
  datePickerSize?: Exclude<typeof SIZES[number], 'tiny' | 'smallest' | 'hero' | 'fullscreen'>;
  /** Whether the picker should automatically open on input focus. True by default. */
  openPickerOnFocus?: boolean;
  /** Whether the input should have button for value clearing. False by default. */
  clearable?: boolean;
  /** Whether user is able to type into the input. True by default. */
  typeable?: IsTypeable;
  /** Error text that is displayed when typed date is invalid. */
  errorText?: string;
}

export type AllowedDisabledDays = Modifier | Date[];
interface DayPickerProps extends Omit<ReactDayPickerProps, 'modifiers' | 'disabledDays'> {
  numberOfMonths?: number;
  showOutsideDays?: boolean;
  showWeekNumbers?: boolean;
  withMonthPicker?: boolean;
  disabledDays?: AllowedDisabledDays;
}

function DatePickerInput<IsTypeable extends boolean = true>({
  className,
  dayPickerProps,
  footer,
  inverse = false,
  inputProps,
  locale = 'en-GB',
  popoverProps,
  size = 'medium',
  inputSize,
  datePickerSize,
  formatDate: customFormatDate,
  openPickerOnFocus = true,
  clearable = false,
  onChange,
  onBlur,
  typeable = true as IsTypeable,
  errorText,
  ...others
}: DatePickerInputProps<IsTypeable>) {
  const getFormattedDateString = (date: Date) => {
    if (!date) {
      return '';
    }

    if (typeable || !customFormatDate) {
      return formatDate(date, DEFAULT_FORMAT, locale);
    }

    return customFormatDate(date, locale);
  };
  const [isPopoverActive, setIsPopoverActive] = useState(false);
  const [popoverAnchorEl, setPopoverAnchorEl] = useState<Element | null>(null);
  const [selectedDate, setSelectedDate] = useState<Date | undefined>(others.selectedDate);
  const [displayError, setDisplayError] = useState(false);
  const [inputValue, setInputValue] = useState(others.selectedDate ? getFormattedDateString(others.selectedDate) : '');
  const handleInputValueChange = (value: string) => {
    setDisplayError(false);
    setInputValue(value);
  };

  useEffect(() => {
    setSelectedDate(others.selectedDate);
    handleInputValueChange(others.selectedDate ? getFormattedDateString(others.selectedDate) : '');
  }, [others.selectedDate]);

  const handleInputFocus = (event: React.FocusEvent<HTMLElement>) => {
    if (inputProps?.readOnly) {
      return;
    }
    if (openPickerOnFocus) {
      setPopoverAnchorEl(event.currentTarget);
      setIsPopoverActive(true);
      inputProps?.onFocus && inputProps.onFocus(event);
    } else {
      inputProps?.onFocus && inputProps.onFocus(event);
    }
  };

  const handleInputClick = (event: React.MouseEvent<HTMLElement>) => {
    if (!openPickerOnFocus) {
      setPopoverAnchorEl(event.currentTarget);
      setIsPopoverActive(true);
      inputProps?.onFocus && inputProps.onFocus(event as unknown as React.FocusEvent<HTMLElement>);
    }
    inputProps?.onClick && inputProps.onClick(event);
  };

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const value = event.target.value;
    const date = parseMultiFormatsDate(value, ALLOWED_DATE_FORMATS, locale);
    if (date && isAllowedDate(date, dayPickerProps?.disabledDays)) {
      setSelectedDate(date);
    }
    handleInputValueChange(value);
  };

  const handleInputBlur = (event: React.FocusEvent<HTMLElement>) => {
    inputProps?.onBlur && inputProps.onBlur(event);
    if (typeable && !customFormatDate && inputValue) {
      const date = parseMultiFormatsDate(inputValue, ALLOWED_DATE_FORMATS, locale);
      if (date && isAllowedDate(date, dayPickerProps?.disabledDays)) {
        handleInputValueChange(getFormattedDateString(date));
        onChange && onChange(date);
      } else {
        setDisplayError(true);
      }
    }
  };

  const handlePopoverClose = () => {
    onBlur && onBlur();
    setIsPopoverActive(false);
  };

  const handleDatePickerDateChange = (date: Date) => {
    setIsPopoverActive(false);
    setSelectedDate(date);
    handleInputValueChange(getFormattedDateString(date));
    onChange && onChange(date);
  };

  const renderIcon = () => {
    return (
      <Icon color={inverse ? 'teal' : 'neutral'} tint={inverse ? 'light' : 'darkest'} marginRight={2}>
        <IconCalendarSmallOutline />
      </Icon>
    );
  };

  const handleClear = (event: MouseEvent) => {
    // Prevents opening datepicker on clicking of this
    event.preventDefault();
    setIsPopoverActive(false);
    setSelectedDate(undefined);
    handleInputValueChange('');
    onChange && onChange(undefined);
  };

  const renderClearIcon = () => {
    return clearable && selectedDate ? (
      <Icon
        color={inverse ? 'teal' : 'neutral'}
        tint={inverse ? 'light' : 'darkest'}
        marginRight={2}
        onClick={handleClear}
      >
        <IconCloseBadgedSmallFilled />
      </Icon>
    ) : null;
  };

  const boxProps = pickBoxProps(others);
  const internalError = displayError ? errorText || true : false;
  return (
    <Box className={className} {...boxProps}>
      <Input
        inverse={inverse}
        prefix={renderIcon()}
        suffix={renderClearIcon()}
        size={inputSize || size}
        width="120px"
        noInputStyling={!typeable}
        {...inputProps}
        error={inputProps?.error || internalError}
        onClick={handleInputClick}
        onFocus={handleInputFocus}
        onBlur={handleInputBlur}
        className={theme['date-picker-input']}
        value={inputValue}
        onChange={handleInputChange}
      />
      <Popover
        active={isPopoverActive}
        anchorEl={popoverAnchorEl}
        backdrop="transparent"
        maxWidth="min-content"
        onEscKeyDown={handlePopoverClose}
        onOverlayClick={handlePopoverClose}
        position="end"
        offsetCorrection={30}
        returnFocusToSource={false}
        withFocusTrap={!typeable}
        {...popoverProps}
      >
        <Box overflowY="auto">
          <DatePicker
            className={theme[`is-${datePickerSize || size}`]}
            onChange={handleDatePickerDateChange}
            selectedDate={selectedDate as Date}
            size={datePickerSize || size}
            {...dayPickerProps}
          />
        </Box>
        {footer && (
          <Box
            backgroundColor="neutral"
            backgroundTint="light"
            borderColor="neutral"
            borderTint="normal"
            borderTopWidth={1}
            flex="1 0 auto"
            overflowX="hidden"
            paddingHorizontal={3}
            paddingVertical={3}
            tabIndex={0}
          >
            {footer}
          </Box>
        )}
      </Popover>
    </Box>
  );
}

export default DatePickerInput;
