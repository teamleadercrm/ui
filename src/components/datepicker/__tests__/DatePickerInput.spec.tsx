import React from 'react';
import { render } from '@testing-library/react';
import DatePickerInput from '../DatePickerInput';

describe('Component - DatePickerInput', () => {
  it('renders', async () => {
    const { asFragment } = render(<DatePickerInput selectedDate={new Date('1965-03-15')} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
