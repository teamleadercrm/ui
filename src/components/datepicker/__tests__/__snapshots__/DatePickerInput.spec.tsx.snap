// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - DatePickerInput renders 1`] = `
<DocumentFragment>
  <div
    class="box"
    data-teamleader-ui="date-picker-input"
  >
    <div
      class="box wrapper"
    >
      <div
        class="input-wrapper"
      >
        <div
          class="input-inner-wrapper"
          style="width: 120px; flex: 0 1 auto;"
        >
          <div
            class="prefix-wrapper"
          >
            <span
              class="box align-items-center display-flex margin-right-2 neutral darkest"
              data-teamleader-ui="icon"
            >
              <svg
                fill="currentColor"
                height="14"
                preserveAspectRatio="xMidYMid meet"
                style="opacity: 0.84; vertical-align: middle;"
                viewBox="0 0 14 14"
                width="14"
              >
                <path
                  clip-rule="evenodd"
                  d="M9 2a1 1 0 00-1 1H6a1 1 0 00-2 0c-1.103 0-2 .897-2 2v5c0 1.103.897 2 2 2h6c1.103 0 2-.897 2-2V5c0-1.103-.897-2-2-2a1 1 0 00-1-1zm1 8H4V6h6v4z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
          <input
            class="input is-medium"
            type="text"
            value="15/03/1965"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
